### ISSUE TITLE
Roles requirements.yml directory and repository structure

### ISSUE THREAD

**Issue by jeffreznik** [ansible-semaphore/semaphore#1609]
Hi there, relatively new to Semaphore and Ansible in general. My git repository contains code other than Ansible playbooks, and is structured so that all Ansible related files are in an `ansible` subdirectory in the root of the project:
```
Repository structure:
/ (root)
├── ansible/
│   ├── group_vars/
│   │   ├── ...
│   ├── playbooks/
│   │   ├── initialize.yml
│   ├── roles/
│   │   ├── requirements.yml
├── module2/
│   ├── ...
├── module3/
│   ├── ...
```
So in Semaphore I create Task Templates pointing to my playbook files like this: `ansible/playbooks/initialize.yml`

My playbook has role requirements listed in `ansible/roles/requirements.yml`. Problem is Semaphore assumes the `roles` folder is in the root of the project. Is there a way to change this behavior? What is the recommended approach for installing roles if your project does not conform to normal convention? It would be nice if Semaphore looked "one directory up" from your playbook file location for the `roles` directory (which would not break existing convention), or if you could specify a subdirectory in your repo where Ansible-related files are located.

I am running Semaphore via Docker (latest version: 2.9.4).

**Comment by jeffreznik** [comment]
For now I've just symlinked `/ansible/roles/` to `/roles/` and it seems to be working. Would still be nice to have a cleaner approach to this.

**Comment by Frickeldave** [comment]
Me too. In my environment semaphore is expecting the roles folder in "<root>/playbooks/roles", but i stored them into "<root>/roles"...

**Comment by rwaffen** [comment]
add a "ansible.cfg" in the root of your repository and add these lines

```
[defaults]
roles_path = roles:/tmp/semaphore/.ansible/roles:/usr/share/ansible/roles
collections_paths = collections:/tmp/semaphore/.ansible/collections:/usr/share/ansible/collections

```

I am also quite new to this, but I read the docu and came up with a template for a new ansible repository, so that i don't have to think about every time: https://github.com/betadots/ansible-control-repo-template

**Comment by jeffreznik** [comment]
`roles_path` is where ansible should look for installed roles. What I'm trying to do is tell Semaphore where my requirements.yml file is so it can install them. Normally you need to call `ansible-galaxy` before running a playbook but Semaphore does this for you.

**Comment by rwaffen** [comment]
when i do it like this, it searches also for requirements.yml there and installs any role from it.

**Comment by jeffreznik** [comment]
You have `roles/requirements.yml` in your project root, my initial question was how do you tell Semaphore to see it when it's not?

**Comment by rwaffen** [comment]
hmm okay, i see your point. i thought this is more a ansible then a semaphore thing. and by telling ansible where my roles are, it should find the yaml. but I didn't test it.

**Comment by riker09** [comment]
@jeffreznik Did you find another solution besides symlinking `/ansible/roles/` to `/roles/`? I have the exact same project setup and I am facing the exact same issue.

**Comment by jeffreznik** [comment]
No, unfortunately this is still the only solution that worked for me.

**Comment by tboerger** [comment]
I think there is an open pull request which sets the role path relative to the playbook.

**Comment by SvartrShell** [comment]
can anyone point me to why i have collections/requirements.yml and roles/requirements.yml in the ROOT of my repo and still sephamore wont detect it at all? has anyone got an example of what the file should look like?

**Comment by robled** [comment]
If you don't want to create messy symlink hacks in your repo, I found another way to do this:

- Variable Groups -> New Group
- On the VARIABLES tab, add a new Extra variable with key `DEFAULT_ROLES_PATH` and value with your subdir, e.g. `ansible/roles`
- Add the Variable Group to your task template
